### PR TITLE
#169817552 add user role settings

### PR DESCRIPTION
**Pull request by higustave-ops**
#### What does this PR do?
`create user role settings endpoint`
#### Description of Task to be completed?
`To create user role settings endpoint feature by allowing a super admin to be able to update/reset other user's roles`
#### How should this be manually tested?
* Clone this repo
* Navigate in the repo root using `CMD` or any `TERMINAL`
* Install all packages using `npm install`
* Type `npm run dev` for Postman testing And `npm run test` for mocha testing
* Open postman choose PATCH `localhost:3000/api/admin/reset-role`
* Click `body` then `x-www-form-urlencoded`
* Fill in `userEmail` and `newRole`
* Click Headers and Type `Authorization` paste in your `After logging in Token` as value
* Click `SEND` on the top right corner of Postman

#### What are the relevant pivotal tracker stories?
- [#169817552](https://www.pivotaltracker.com/story/show/169817552)

#### Screenshots (if appropriate)
`Postman screenshot`
Screen Shot 2020-01-08 at 15 50 44

### CODE REVIEW

File: src/tests/fixtures/users.fixture.js [hunk 1]
```diff
@@ -127,6 +152,7 @@ export const createUsers = async () => {
   await Users.create(createUser);
   await Users.create({ ...loggedInUser, token: loggedInToken });
   await Users.create({ ...userWithNoTrip, token: userWithNoTripToken });
+  await Users.create({ ...superAdminTestData, token: superAdminToken });
 };
 export const cleanDb = async () => {
```
Comment (erickyvand) [inline src/tests/fixtures/users.fixture.js]: Hi @higustave-ops good work so far, but I wanted to know what did you create cleanDb to delete table while we have that function in createUsers functions?

File: src/controllers/user-role-settings.controller.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  * super_admin can assign a role to user by email
+  * @static
+  * @description PATCH /api/admin/reset-user-role
+  * @param {object} req request object
+  * @param {object} res response object
+  * @memberof UserRoleSettingsController
+  * @returns {object} ResponseService
+  */
+  static async changeUserRole(req, res) {
+    const userToken = JwtService.verifyToken(req.headers.authorization);
```
Comment (adesege) [inline src/controllers/user-role-settings.controller.js]: Why are you verifying the token here? There is a middleware that does that already.

File: src/controllers/user-role-settings.controller.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    const userData = await UserService.findUserByProperty({ email: userToken.email });
+    if (userData.role === 'super_admin') {
```
Comment (adesege) [inline src/controllers/user-role-settings.controller.js]: This check should be moved to a middleware

File: src/controllers/user-role-settings.controller.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      const { userEmail, newRole } = req.body;
+      const userExist = await UserService.findUserByProperty({ email: userEmail.trim() });
+      if (userExist && userExist.role !== 'super_admin') {
```
Comment (adesege) [inline src/controllers/user-role-settings.controller.js]: This if statement is unnecessary here because `userExist.role` will always be a `super_admin` here.

File: src/controllers/user-role-settings.controller.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        if (userExist.role === newRole.trim()) {
```
Comment (adesege) [inline src/controllers/user-role-settings.controller.js]: If the role is the same, you shouldn't throw an error. Just return a success message instead

File: src/controllers/user-role-settings.controller.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+          ResponseService.setError(400, `User Role Not Updated. ${userExist.email} is already a ${userExist.role}`);
+          ResponseService.send(res);
+        } else {
+          await UserService.updateUser({ email: userEmail.trim() }, { role: newRole.trim() });
```
Comment (adesege) [inline src/controllers/user-role-settings.controller.js]: You should use the user's id which is a primary id in the where object instead of email

File: src/controllers/user-role-settings.controller.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+          const data = await UserService.findUserByProperty({ email: userEmail.trim() });
+          const { password, token, createdAt, updatedAt, isVerified, ...datas } = data.dataValues;
+          ResponseService.setSuccess(200, 'User role successfully updated', datas);
```
Comment (adesege) [inline src/controllers/user-role-settings.controller.js]: Since you updated only the user's role, you should only send that as a response.

File: src/routes/user-role-settings.route.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import express from 'express';
+import authMiddleware from '../middlewares/auth.middleware';
+import UserRoleSettingsController from '../controllers/user-role-settings.controller';
+import validateUserRole from '../validations/user-role-settings.validation';
+
+const router = express.Router();
+
+router.patch('/reset-user-role', authMiddleware.checkUserLoggedIn, validateUserRole, UserRoleSettingsController.changeUserRole);
```
Comment (adesege) [inline src/routes/user-role-settings.route.js]: The path can be `reset-role` instead

File: src/seeders/20200103073111-users-seed-file.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+ * @class
+ * @param {object} queryInterface
+ * @param {object} Sequelize
+ */
+export function up(queryInterface) {
+  return queryInterface.bulkInsert('Users', [{
+    firstName: 'James',
+    lastName: 'Bond',
+    email: 'jamesbond@gmail.com',
+    password: process.env.SUPER_ADMIN_PASSWORD,
```
Comment (adesege) [inline src/seeders/20200103073111-users-seed-file.js]: Why are you passing the password as an environment variable? All data should be auto-generated using faker

File: src/tests/fixtures/users.fixture.js [hunk 10]
```diff
@@ -78,6 +78,31 @@ export const createUser = {
   createdAt: new Date(),
   updatedAt: new Date(),
 };
+export const superAdminTestData = {
+  id: 1,
```
Comment (adesege) [inline src/tests/fixtures/users.fixture.js]: Use Faker to auto-generate your fixture data

File: src/validations/user-role-settings.validation.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  userEmail: Joi.string()
+    .email({ minDomainSegments: 2 })
+    .required()
+    .trim()
+    .messages({
+      'string.empty': 'Please enter email',
+      'string.email': 'Enter valid email i.e: email@example.com',
+      'any.required': 'Email is required'
+    }),
+  newRole: Joi.string()
```
Comment (adesege) [inline src/validations/user-role-settings.validation.js]: Use a more descriptive variable name

**Review by adesege — CHANGES_REQUESTED** [review]

File: src/tests/fixtures/users.fixture.js (hunk 1 above)
Comment (higustave-ops) [reply src/tests/fixtures/users.fixture.js]: @adesege and @erickyvand 

* For the signup test need a fresh database table so the `cleanDb` function on `line 157` will do the job to clean it for signup test. 

* For `createUsers` function, every test may create its users. Thus a need to delete other users and create its own. That's the reason why we have both functions

File: src/controllers/user-role-settings.controller.js (hunk 4 above)
Comment (higustave-ops) [reply src/controllers/user-role-settings.controller.js]: @adesege Here the user role can not only be `super_admin` as i am getting the email from `req.body`  `line 24` and such email may be associated to any user role type. And the pupose of that line `26` is: if the user exists and he (that user) is not `super_admin` he (that user) can be updated. A `super_admin` can not be updated.

File: src/controllers/user-role-settings.controller.js (hunk 6 above)
Comment (higustave-ops) [reply src/controllers/user-role-settings.controller.js]: Pivotal tracker story says `update user role by email`. Email is also unique. in the DB

File: src/seeders/20200103073111-users-seed-file.js (hunk 9 above)
Comment (higustave-ops) [reply src/seeders/20200103073111-users-seed-file.js]: @adesege  Here its in the seed file. Once a dummy seed admin user is inserted, i will need to log him in and get the token that will allow him to use `/reset-role` route. if i use faker, i will not be able to know his `email` and `password`. I only used faker for the fixtures in test

Comment (adesege) [reply src/seeders/20200103073111-users-seed-file.js]: Then you should create the user before the test runs. Check other tests that use the before hook

File: src/tests/fixtures/users.fixture.js [hunk 12]
```diff
@@ -79,6 +79,38 @@ export const createUser = {
   updatedAt: new Date(),
 };
 
+const adminRole = 'super_admin';
```
Comment (higustave-ops) [inline src/tests/fixtures/users.fixture.js]: @adesege here  `line 82~86` i need specific data and faker may generate different data which will cause unexpected results/behavior of the test. Th reason why i excluded them from being generated by faker.

Comment (adesege) [reply src/tests/fixtures/users.fixture.js]: That's fine. However, Fakerjs should be able to generate a random language and currency. That way, you are able to test for other language and currency which are not necessarily English and Euro.

File: src/seeders/20200103073111-users-seed-file.js (hunk 9 above)
Comment (higustave-ops) [reply src/seeders/20200103073111-users-seed-file.js]: @adesege This user here is not for mocha test. its for dev environment for `POSTMAN`.
For `mocha test`, i have generated a faker admin user data except few fields.

File: src/tests/fixtures/users.fixture.js (hunk 12 above)
Comment (higustave-ops) [reply src/tests/fixtures/users.fixture.js]: @adesege Yes, fakerjs can generate language and current but for each test faker will generate different data for those above fields and i will not be able to recognise a new regenerated value. Mean for each time we run test i will re write the test for those regenerated values which may also not pass because i will be writing test of  unknown output.

File: src/seeders/20200103073111-users-seed-file.js (hunk 9 above)
Comment (adesege) [reply src/seeders/20200103073111-users-seed-file.js]: I would rather have you manually create a user and assign the super_admin role to it than using a seeder file to do that. This have some security implications to it. This seeder file can be run in production and your password hash should not leave your database.

File: src/tests/fixtures/users.fixture.js (hunk 12 above)
Comment (adesege) [reply src/tests/fixtures/users.fixture.js]: Then you are writing your test the wrong way. Your tests should be able to run in isolation. If the DB is refreshed, the tests should still pass.

The reason why you write test is to take care of different edge cases that may arise when users use your platform.

Does Barefoot Nomad accept other currencies and languages aside from Euro and English?

File: src/seeders/20200103073111-users-seed-file.js (hunk 9 above)
Comment (higustave-ops) [reply src/seeders/20200103073111-users-seed-file.js]: @adesege Ok, then i am gonna prevent the application to create this admin user in dev environment and add this user manually in the DB.

File: src/tests/fixtures/users.fixture.js (hunk 12 above)
Comment (higustave-ops) [reply src/tests/fixtures/users.fixture.js]: @adesege i have chosen to only use `Euro, Pound, Dollar` for currencies and `english, french` for languages Just for preventing users to input whatever language and currencies they want.

Comment (adesege) [reply src/tests/fixtures/users.fixture.js]: Then your tests should take into consideration other currencies and languages as well. If Fakerjs doesn't have the data you want, you can create an array of possible valid values and use the `Faker#randomize` method to pick any of the items randomly.

At the route level, you should also use Joi to check that the user is setting the right value.

Comment (higustave-ops) [reply src/tests/fixtures/users.fixture.js]: @adesege  Ok, understood. That is a better solution. I am gonna implement it. I will let you know once done. Thank you

File: src/routes/user-role-settings.route.js [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import express from 'express';
+import authMiddleware from '../middlewares/auth.middleware';
+import UserRoleSettingsController from '../controllers/user-role-settings.controller';
+import validateUserRole from '../validations/user-role-settings.validation';
+
+const router = express.Router();
+
+router.patch('/reset-roles', authMiddleware.checkUserLoggedIn, validateUserRole, authMiddleware.checkUserRole, UserRoleSettingsController.changeUserRole);
```
Comment (higustave-ops) [inline src/routes/user-role-settings.route.js]: @adesege  here i have changed from `reset-role` to `reset-roles` as you required.

File: src/middlewares/auth.middleware.js [hunk 14]
```diff
@@ -32,6 +32,16 @@ const authMiddleware = {
       ResponseService.setError(401, 'No Token supplied');
       return ResponseService.send(res);
     }
+  },
+
+  checkUserRole: async (req, res, next) => {
```
Comment (adesege) [inline src/middlewares/auth.middleware.js]: This function can be renamed to reflect the right role it is checking for. `checkUserRole` can be confused to mean you are checking the role of a `user` and not a `super admin`

File: src/tests/fixtures/users.fixture.js [hunk 15]
```diff
@@ -15,9 +15,14 @@ export const signupFixtures = {
 
 export const signedUpUserToken = JwtService.generateToken({ email: signupFixtures.email });
 export const unregisteredEmail = {
-  email: 'higustave123@gmail.com'
+  email: 'higustave123346432@gmail.com'
```
Comment (adesege) [inline src/tests/fixtures/users.fixture.js]: Always use randomly generated value. I've said this a couple of times.

File: src/validations/user-role-settings.validation.js [hunk 16]
```diff
@@ -0,0 +1,43 @@
+import Joi from '@hapi/joi';
```
Comment (adesege) [inline src/validations/user-role-settings.validation.js]: https://github.com/andela/the_spinners-backend/pull/42#discussion_r364883974

File: src/validations/user-role-settings.validation.js [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+      'any.required': 'userRole is required'
+    })
+}).options({ abortEarly: false });
+
+const validateHandler = (schema, body, res, next) => {
+  const { error } = schema.validate(body);
+  if (error) {
+    const errors = [];
+    const { details } = error;
+    details.forEach(({ message }) => {
```
Comment (adesege) [inline src/validations/user-role-settings.validation.js]: Always use .map

File: src/validations/user-role-settings.validation.js (hunk 16 above)
Comment (higustave-ops) [reply src/validations/user-role-settings.validation.js]: @adesege here imported `Joi` and being used for validation in this file. #42 imported `import joiDate from '@hapi/joi-date'` and did not use it. I think there is difference between our imports. Mine is used. Thank you

File: src/tests/fixtures/users.fixture.js (hunk 15 above)
Comment (higustave-ops) [reply src/tests/fixtures/users.fixture.js]: Sorry for unimplementation. Will not happen again.

File: package.json [hunk 18]
```diff
@@ -17,12 +17,12 @@
...
     "migrate:drop": "sequelize db:migrate:undo:all",
+    "seed:create": "npx sequelize-cli db:seed:all",
     "dev": "nodemon --exec babel-node src/server.js",
     "coverage": "nyc report --reporter=text-lcov | coveralls",
-    "pretest": "npm run migrate:drop && npm run migrate:up && npm run seed:up",
+    "pretest": "npm run migrate:drop && npm run migrate:up",
     "test": "export NODE_ENV=test && nyc --reporter=html --reporter=text mocha  --require  @babel/register src/tests/**/*.test.js --timeout 100000 --exit",
-    "migrate:up": "npx sequelize-cli db:migrate",
-    "seed:up": "npx sequelize-cli db:seed:all"
+    "migrate:up": "npx sequelize-cli db:migrate && npm run seed:create"
```
Comment (adesege) [inline package.json]: Sometimes when you run migrations, you don't want to run seeders too.  Remember, dummy data is not allowed in production.

Comment (adesege) [reply package.json]: If you need to run seeders in ci, add the script to travis config instead

File: src/controllers/user-role-settings.controller.js [hunk 19]
```diff
@@ -0,0 +1,34 @@
+import ResponseService from '../services/response.service';
+import UserService from '../services/user.service';
+/**
+ *
+ *
+ * @class UserRoleSettingsController
+ */
+class UserRoleSettingsController {
```
Comment (adesege) [inline src/controllers/user-role-settings.controller.js]: Change this class to `User` or `Settings` controller. Gave feedback that files should not be created in isolation but they must be seen as a group or collection.

File: package.json (hunk 18 above)
Comment (higustave-ops) [reply package.json]: @adesege Ok solved now

File: src/controllers/user-role-settings.controller.js (hunk 19 above)
Comment (higustave-ops) [reply src/controllers/user-role-settings.controller.js]: @adesege  Ok solved now on by renaming all files related to this feature in both `controllers` and `swagger` and  `routes`

File: src/models/users.js [hunk 20]
```diff
@@ -13,9 +13,8 @@ export default (sequelize, DataTypes) => {
     lineManagerId: DataTypes.INTEGER,
     role: DataTypes.ENUM('super_admin', 'travel_admin', 'travel_team_member', 'manager', 'requester'),
     profilePicture: DataTypes.STRING(1234),
-    token: DataTypes.STRING,
-    isVerified: DataTypes.BOOLEAN,
-    lineManagerId: DataTypes.INTEGER
```
Comment (adesege) [inline src/models/users.js]: Why is `lineManagerId` removed? I see it is used by other team members.

Comment (higustave-ops) [reply src/models/users.js]: @adesege  It was duplicated. You can see `Line 13` and `Line 18`

File: src/routes/index.js [hunk 21]
```diff
@@ -1,14 +1,14 @@
...
 import accommodationRoute from './accommodation.route';
+import SettingsRoute from './settings.route';
 
 const app = express();
 
 app.use('/api/auth', authRoute);
 app.use('/api/', tripRoute);
-app.use('/api/', profileSettingsRoute);
 app.use('/api/accommodations', accommodationRoute);
+app.use('/api/users', SettingsRoute);
```
Comment (adesege) [inline src/routes/index.js]: Use `/api/users/settings` as the route context here

File: src/tests/fixtures/users.fixture.js [hunk 22]
```diff
@@ -59,6 +59,24 @@ export const loggedInUser = {
   updatedAt: new Date(),
 };
 
+export const unverifiedUser = {
+  id: 80,
```
Comment (adesege) [inline src/tests/fixtures/users.fixture.js]: Is there a reason why this id is not auto-generated?

File: src/routes/index.js (hunk 21 above)
Comment (higustave-ops) [reply src/routes/index.js]: @adesege Ok solved

File: src/tests/fixtures/users.fixture.js (hunk 22 above)
Comment (higustave-ops) [reply src/tests/fixtures/users.fixture.js]: @adesege Ok solved on my faker data but left it intact on other team mates faker data. They said that they have implemented it in theirs.

**Review by adesege — APPROVED** [review]

**Comment by adesege** [comment]
Your travis build has failed @higustave-ops

**Comment by higustave-ops** [comment]
> Your travis build has failed @higustave-ops

@adesege Now is passing, some tests for trips were failing. Had to rebuild it. Thank you

**Comment by higustave-ops** [comment]
> > Your travis build has failed @higustave-ops
> 
> @adesege Now is passing, some tests for trips were failing. Had to rebuild it. Thank you

I am going to rebase and let you know once finished

File: src/tests/settings/settings.test.js [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+        res.body.should.be.an('object');
+        res.should.have.status(403);
+        res.body.should.have.property('message').equal('Only super admin can reset user role');
+        done();
+      });
+  });
+
+  it('It should NOT allow Super-Admin user to reset user role: User Not Found', (done) => {
+    const newEmail = 'paul@gmail.com';
+    newRoleData.userEmail = 'paul@gmail.com';
```
Comment (victor-abz) [inline src/tests/settings/settings.test.js]: use faker for this data

File: package.json [hunk 24]
```diff
@@ -17,12 +17,12 @@
     "build": "babel src  --out-dir production",
     "start": "node production/server.js",
     "migrate:drop": "sequelize db:migrate:undo:all",
+    "seed:create": "npx sequelize-cli db:seed:all",
     "dev": "nodemon --exec babel-node src/server.js",
     "coverage": "nyc report --reporter=text-lcov | coveralls",
-    "pretest": "npm run migrate:drop && npm run migrate:up && npm run seed:up",
+    "pretest": "npm run migrate:drop && npm run migrate:up && npm run seed:create",
```
Comment (victor-abz) [inline package.json]: Remove the seed command as you are having no seeds

File: package.json [hunk 25]
```diff
@@ -17,12 +17,12 @@
     "build": "babel src  --out-dir production",
     "start": "node production/server.js",
     "migrate:drop": "sequelize db:migrate:undo:all",
+    "seed:create": "npx sequelize-cli db:seed:all",
```
Comment (victor-abz) [inline package.json]: Remove the seed command as you are having no seeds

File: src/routes/index.js [hunk 26]
```diff
@@ -1,14 +1,14 @@
...
 import accommodationRoute from './accommodation.route';
+import SettingsRoute from './settings.route';
 
 const app = express();
 
 app.use('/api/auth', authRoute);
 app.use('/api/', tripRoute);
-app.use('/api/', profileSettingsRoute);
 app.use('/api/accommodations', accommodationRoute);
+app.use('/api/users/settings', SettingsRoute);
```
Comment (victor-abz) [inline src/routes/index.js]: variable name should be camelCase as the file is not a class

File: src/controllers/settings.controller.js [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+  * @static
+  * @description PATCH /api/admin/reset-user-role
+  * @param {object} req request object
+  * @param {object} res response object
+  * @memberof UserRoleSettingsController
+  * @returns {object} ResponseService
+  */
+  static async changeUserRole(req, res) {
+    const { userEmail, userRole } = req.body;
+    const userExist = await UserService.findUserByProperty({ email: userEmail.trim() });
```
Comment (erickyvand) [inline src/controllers/settings.controller.js]: Hey @higustave-ops, if you rename `userExist` to `findUser` or `findUserRole` or `user` could be better, because I can see down you are using this variable to find a role of a user. what do you think about that?

File: src/tests/settings/settings.test.js [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+      .end((err, res) => {
+        res.body.should.be.an('object');
+        res.should.have.status(403);
+        res.body.should.have.property('message').equal('Only super admin can reset user role');
+        done();
+      });
+  });
+
+  it('It should NOT allow Super-Admin user to reset user role: User Not Found', (done) => {
+    const newEmail = 'paul@gmail.com';
```
Comment (erickyvand) [inline src/tests/settings/settings.test.js]: can you put this in separate file the call it here?

File: src/tests/settings/settings.test.js [hunk 29]
```diff
@@ -0,0 +1,75 @@
...
+        res.body.should.be.an('object');
+        res.should.have.status(404);
+        res.body.should.have.property('message').equal(`User role not updated. ${newEmail} is either a super admin or Not Registered`);
+        done();
+      });
+  });
+
+  it('It should NOT allow Super-Admin user to reset user role: Incomplete input data', (done) => {
+    const { userRole } = newRoleData;
+    const newData = { userRole };
```
Comment (erickyvand) [inline src/tests/settings/settings.test.js]: I think you can remove this too and put in a separate file

File: src/tests/settings/settings.test.js (hunk 28 above)
Comment (higustave-ops) [reply src/tests/settings/settings.test.js]: I did it in victor's feedback

File: src/tests/settings/settings.test.js (hunk 29 above)
Comment (higustave-ops) [reply src/tests/settings/settings.test.js]: Here i am destructuring to only get `userRole` field from `newRoleData` object.

File: src/controllers/settings.controller.js (hunk 27 above)
Comment (higustave-ops) [reply src/controllers/settings.controller.js]: @erickyvand , @adesege said that for a variable that will return a boolean, we use such kind of names to reflect to the return value. example: `isUserAdmin`: true or false `userExist`: true or false.
On that line i was checking whether a user exists in the DB. It can return `true` or `false`.

File: package.json (hunk 25 above)
Comment (higustave-ops) [reply package.json]: I have rebase after Gad's merge. And we discussed and told me that he will need it. Once i remove them, his test that i have are failing

**Review by victor-abz — APPROVED** [review]
LGTM

**Review by erickyvand — APPROVED** [review]

File: src/controllers/settings.controller.js [hunk 30]
```diff
@@ -0,0 +1,34 @@
...
+    if (userExist && userExist.role !== 'super_admin') {
+      await UserService.updateUser({ email: userEmail.trim() }, { role: userRole.trim() });
+      const data = await UserService.findUserByProperty({ email: userEmail.trim() });
+      const { role } = data;
+      const responseData = { role };
```
Comment (gadishimwe) [inline src/controllers/settings.controller.js]: This is an extra line which is not necessary, you should pass that as an object in setSuccess() method instead.

File: package.json (hunk 25 above)
Comment (adesege) [reply package.json]: Then you shouldn't modify the script. Develop already has a script to run seeders.

File: src/controllers/settings.controller.js (hunk 27 above)
Comment (adesege) [reply src/controllers/settings.controller.js]: `findUserByProperty` does not return a boolean but an object if it can find a user or null if it can't. You should take @erickyvand feedback into consideration.

File: src/validations/role.validation.js [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+      'any.required': 'userRole is required'
+    })
+}).options({ abortEarly: false });
+
+const validateHandler = (schema, body, res, next) => {
+  const { error } = schema.validate(body);
+  if (error) {
+    const errors = [];
+    const { details } = error;
+    details.map(({ message }) => errors.push((message.replace(/[^a-zA-Z0-9 .-]/g, ''))));
```
Comment (adesege) [inline src/validations/role.validation.js]: Refactor this line accordingly. This is the umpteenth time I will be giving feedback on how to use `.map`

File: src/middlewares/auth.middleware.js [hunk 32]
```diff
@@ -32,6 +32,16 @@ const authMiddleware = {
       ResponseService.setError(401, 'No Token supplied');
       return ResponseService.send(res);
     }
+  },
+
+  verifyAdmin: async (req, res, next) => {
```
Comment (gadishimwe) [inline src/middlewares/auth.middleware.js]: This method name is not descriptive enough and this is similar to feedback that was given on PR #42 , please rename it.

File: package.json (hunk 25 above)
Comment (higustave-ops) [reply package.json]: Sorry for modifying.Will not happen again

File: src/controllers/settings.controller.js (hunk 27 above)
Comment (higustave-ops) [reply src/controllers/settings.controller.js]: @adesege yes. I now changed

File: src/middlewares/auth.middleware.js (hunk 32 above)
Comment (higustave-ops) [reply src/middlewares/auth.middleware.js]: I have implemented according @adesege requirements. @gadishimwe any suggestion of how you want me to rename it?

Comment (gadishimwe) [reply src/middlewares/auth.middleware.js]: I can suggest verifyIfUserIsAdmin

Comment (higustave-ops) [reply src/middlewares/auth.middleware.js]: @gadishimwe Yes applied

**Comment by gadishimwe** [comment]
@higustave-ops Why is your coverage dropping?

**Comment by higustave-ops** [comment]
> @higustave-ops Why is your coverage dropping?

@gadishimwe because of Victor's work on notifications. His PR doesn't cover 100%. So, after rebasing develop since he merged all PRs will no longer get 100%

**Comment by gadishimwe** [comment]
> > @higustave-ops Why is your coverage dropping?
> 
> @gadishimwe because of Victor's work on notifications. His PR doesn't cover 100%. So, after rebasing develop since he merged all PRs will no longer get 100%

Oh got it.

**Review by gadishimwe — APPROVED** [review]
LGTM
